homeassistant/components/matter/media_player.py

```diff
@@ -0,0 +1,141 @@
+            clusters.LevelControl.Commands.MoveToLevel(level=matter_volume_level)
+        )
+
+    async def async_mute_volume(self, mute: bool) -> None:
+        """Mute or unmute the media player."""
+        # OnOff attribute == True state means volume is on, so HA should show mute switch as off
+        if mute:
+            # Save current volume before muting to be able to restore it
+            current_volume = self._attr_volume_level or 1
+            matter_volume_level = int(current_volume * 254)
+            await self.write_attribute(
+                matter_volume_level, clusters.LevelControl.Attributes.OnLevel
+            )
+            # Send Matter Off command
+            await self.send_device_command(
+                clusters.OnOff.Commands.Off(),
+            )
+        # OnOff attribute == False means volume is off (muted), so HA should show mute switch as on
+        else:
+            await self.send_device_command(
+                clusters.OnOff.Commands.On(),
+            )
+
+    @callback
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        self._calculate_features()
+        matter_volume = self.get_matter_attribute_value(
+            clusters.LevelControl.Attributes.CurrentLevel
+        )
+        LOGGER.debug(
+            "matter_volume: %f for %s",
+            matter_volume,
+            self.entity_id,
+        )
+        if matter_volume == 0:
+            self._attr_is_volume_muted = True
+        # Convert Matter CurrentLevel (0-254) to HA volume level (0-1)
+        else:
+            self._attr_is_volume_muted = False
+            self._attr_volume_level = matter_volume / 254.0
+        # No state in the Speaker endpoint as it dedicated to volume control
+        self._attr_state = MediaPlayerState.ON
+
+    @callback
+    def _calculate_features(self) -> None:
+        """Calculate features for HA MediaPlayer platform."""
+        supported_features = (
+            MediaPlayerEntityFeature.VOLUME_SET
+            | MediaPlayerEntityFeature.VOLUME_MUTE
+            | MediaPlayerEntityFeature.VOLUME_STEP
+            # | MediaPlayerEntityFeature.TURN_ON    # Uncomment if you want to support turning on
+            # | MediaPlayerEntityFeature.TURN_OFF   # Uncomment if you want to support turning off
+        )
+        self._attr_supported_features = supported_features
+        self._attr_volume_step = 0.01  # Matter uses 1-254, so step can be small
+
+
+# Discovery schema(s) to map Matter Attributes to HA entities
+DISCOVERY_SCHEMAS = [
+    MatterDiscoverySchema(
+        platform=Platform.MEDIA_PLAYER,
+        entity_description=MediaPlayerEntityDescription(
+            key="MatterMediaPlayer", name=None
+        ),
+        entity_class=MatterMediaPlayer,
+        required_attributes=(
+            clusters.OnOff.Attributes.OnOff,
+            clusters.LevelControl.Attributes.CurrentLevel,
+            clusters.LevelControl.Attributes.OnLevel,
```

Comment:
After clarifications OnLevel should only be "NULL" for a speaker because it makes no sense, so better remove here

Reply:
When you mute you do not change the volume itself. means: 
* Mute is done via the OnOff endpoint
* changing level changes the volume, but if muted it might not have an effect ( or depending then on the OnOff Feature of Levelcontrol if there is a cluster-internal logic

OnLevel on a speak shall only be NULL and twith this the matter logic preserves "currentLevel"

Reply:
tests with which devices? potentially the devices needs to be fixed :-)